Enabling ILB/ELB on windows using per-node, per-network LB endpoint.

pradipd opened the pull request:
Note: This is 1 of 3 PRs (the other 2 are in libnetwork and swarmkit repos).

Signed-off-by: Pradip Dhara <pradipd@microsoft.com>

**- What I did**
Enabled ILB/ELB (routing mesh) functionality for windows.

**- How I did it**
This change enables ILB/ELB functionality on windows using a per-node, per-network LB endpoint.  The per-node, per-network LB endpoint was discussed with docker 
 as a solution to https://github.com/moby/moby/issues/30820.  Note that only windows is using the LB endpoint.  We have not changed the linux networking code to take advantage of the LB endpoint.


**- How to verify it**
make test on linux (note: I'm still seeing 2 tests occasionally fail.  But, I'm fairly certain they are unrelated to my change.  I'm sending this out now and will continue to debug).
We have added a bunch of powershell scripts to test out functionality. However, they are not public.  We will work with docker folks in figuring out how to make them public.  Ideally, we should get the docker_cli_swarm_tests running on windows, but, that is much bigger task and requires changes in windows.  

**- Description for the changelog**
Enabling ILB/ELB on windows using per-node, per-network LB endpoint.

**- A picture of a cute animal (not mandatory but encouraged)**
![tubby shaved](https://user-images.githubusercontent.com/14915188/29848109-93c9da6a-8cd3-11e7-84c2-7379cbd94fd3.jpg)

pradipd committed:
Enabling ILB/ELB on windows using per-node, per-network LB endpoint.

Signed-off-by: Pradip Dhara <pradipd@microsoft.com>

pradipd commented:
libnetwork PR: https://github.com/docker/libnetwork/pull/1925
swarmkit PR: https://github.com/docker/swarmkit/pull/2363

pradipd commented:
@mavenugo @fcrisciani @msabansal

vieux commented:
@pradipd can you please fix the tests ?

dnephin commented on daemon/network.go:
```diff
@@ -360,6 +387,34 @@ func (daemon *Daemon) createNetwork(create types.NetworkCreateRequest, id string
 	}
 	daemon.LogNetworkEvent(n, "create")
 
+	if agent && !n.Info().Ingress() && n.Type() == "overlay" {
```
Please move this code (and the code in deleteNetwork) to a new function. A separate function will make it easier to unit test, and help the reader understand the different steps performed as part of `createNetwork` and `deleteNetwork`.

pradipd replied:
Done.

dnephin commented on daemon/network.go:
```diff
@@ -360,6 +387,34 @@ func (daemon *Daemon) createNetwork(create types.NetworkCreateRequest, id string
+		sb, err := c.NewSandbox(sandboxName)
+		if err != nil {
+			logrus.Errorf("Failed creating %s sandbox: %v", sandboxName, err)
+			return nil, err
```
I don't think the error needs to be logged.  The conventional way of handling errors is to return them with `errors.Wrapf(err, "failed creating %s sandbox", sandboxName)`.

The errors should end up in the logs.

pradipd commented:
Thanks for the feedback.  I'm addressing it now.  I had to fix some changes in libnetwork PR (https://github.com/docker/libnetwork/pull/1925) and swarmkit PR (https://github.com/docker/swarmkit/pull/2363)

pradipd commented on daemon/network.go:
```diff
@@ -283,7 +288,40 @@ func (daemon *Daemon) CreateNetwork(create types.NetworkCreateRequest) (*types.N
+	endpointName := prefix + "-endpoint"
+	ep, err := n.CreateEndpoint(endpointName, libnetwork.CreateOptionIpam(ip, nil, nil, nil), libnetwork.CreateOptionLoadBalancer())
+	if err != nil {
+		logrus.Errorf("Failed creating %s in sandbox %s: %v", endpointName, sandboxName, err)
```
Note: I left the logrus error messages here, because they were here before for when we were creating the ingress network.

pradipd commented:
@vieux  In order to get the tests passing I need changes from https://github.com/docker/libnetwork/pull/1925 and https://github.com/docker/swarmkit/pull/2363
Do I have to wait for those 2 pull requests to merge?
Or do i vendor in those changes and point vendor.conf to my github repo?

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "windows_routingmesh" git@github.com:pradipd/moby.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842354127368
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

pradipd commented:
@vieux : Can you help me understand the integration tests?
Originally I added a test to docker_cli_swarm_test.go, but that gave a janky error.
integration-cli is deprecated. Please add an API integration test to
17:19:20 ./integration/COMPONENT/. See ./TESTING.md for more details

So, I moved the test to docker_api_swarm_test.go, but the test is exactly the same.  That fixed the Janky error, but, I don't think that was the proper way to fix the test.

I also tried to add the test to integration/service, but, I can't figure out how to run the tests.
How to I run the tests under integration/service?

dnephin commented on integration-cli/docker_api_swarm_test.go:
```diff
@@ -1031,3 +1031,35 @@ func (s *DockerSwarmSuite) TestAPINetworkInspectWithScope(c *check.C) {
+        d := s.AddDaemon(c, true, true)
+
+        // Create a new overlay network
+        out, err := d.Cmd("network", "create", "-d", "overlay", "new-overlay")
```
The API suite should not use `d.Cmd()` please use `testEnv.APIClient()` to get a client, and make calls to the API.

pradipd replied:
Thanks for the clarification.  That is what I thought, but, I saw other tests in the .Cmd() and hence the confusion.  I will clean this up.

dnephin reviewed (commented):
> How to I run the tests under integration/service?

The same way you would the cli suite, `make test-integration`

vieux commented:
ping @tiborvass @fcrisciani @vdemeester can you take a look ?

thaJeztah commented:
hm, not sure I've seen this one before; https://jenkins.dockerproject.org/job/Docker-PRs-s390x/5702/console

```
16:24:32 Running /go/src/github.com/docker/docker/integration/container
16:24:32 INFO: Testing against a local daemon
16:24:32 === RUN   TestCreateFailsWhenIdentifierDoesNotExist
16:24:41 --- FAIL: TestCreateFailsWhenIdentifierDoesNotExist (9.55s)
16:24:41 	protect.go:75: Failed to load frozen images: ApplyLayer exit status 1 stdout:  stderr: unexpected EOF
16:24:41 === RUN   TestCreateWithInvalidEnv
16:24:51 --- FAIL: TestCreateWithInvalidEnv (9.54s)
16:24:51 	protect.go:75: Failed to load frozen images: ApplyLayer exit status 1 stdout:  stderr: unexpected EOF
16:24:51 FAIL
```

I can restart to see if it was just a one-off

vieux commented:
LGTM

dnephin commented on daemon/daemon.go:
```diff
@@ -121,6 +121,9 @@ type Daemon struct {
 	startupDone      chan struct{}
+
+	//Stores The LB IP address for the node on a given a network id.
+	networkToNodeLBIP map[string]net.IP
```
I think this would be more appropriate as a new type (something like LBAttachmentStore?), that way we don't add additional methods onto `Daemon` which is already terribly bloated. The new methods would be on the new type.

Unless there's already a type on daemon that could be used?

Maybe something that wraps both this and netController?

pradipd replied:
What package would you suggest adding LBAttachmentStore to?
We can add it to daemon, but, then we need to import daemon into backend.go and that sort of feels like it violates some abstraction.
I can put it in its own package.
I can also put it in the network package under (daemon/settings.go).  I feel like this is the most appropriate spot.
Other suggestions?

pradipd replied:
I put it in the network package under daemon/settings.go.

dnephin commented on daemon/network.go:
```diff
@@ -283,7 +288,40 @@ func (daemon *Daemon) CreateNetwork(create types.NetworkCreateRequest) (*types.N
+
+	if err := ep.Join(sb, nil); err != nil {
+		logrus.Errorf("Failed joining %s to sanbox %s: %v", endpointName, sandboxName, err)
+		return err
```
These are still going to cause duplicate errors, it should use `errors.Wrapf()` to provide the extra messages with the error.

thaJeztah replied:
Also: typo s/sanbox/sandbox/

pradipd replied:
Fixed.

dnephin commented on integration/service/create_test.go:
```diff
@@ -0,0 +1,113 @@
+				Replicas: &instances,
+			},
+		},
+	}
```
Please move all this setup code out into a function which has a name that describes the setup.

pradipd replied:
Fixed in latest PR.

dnephin commented on integration/service/create_test.go:
```diff
@@ -0,0 +1,113 @@
+	"golang.org/x/net/context"
+)
+
+func TestSwarmLBSandbox(t *testing.T) {
```
The test name should match the endpoint that is being tested, so something like `TestCreateWithLBSandbox`

pradipd replied:
Fixed

dnephin commented on integration/service/create_test.go:
```diff
@@ -0,0 +1,113 @@
+	require.NoError(t, err)
+
+	serviceID := serviceResp.ID
+	poll.WaitOn(t, func(log poll.LogT) poll.Result {
```
Please make this a named function (not a closure) so that the test reads better.

pradipd replied:
Fixed in latest PR.

dnephin commented on integration/service/create_test.go:
```diff
@@ -0,0 +1,113 @@
+	err = client.ServiceRemove(context.Background(), serviceID)
+	require.NoError(t, err)
+
+	poll.WaitOn(t, func(log poll.LogT) poll.Result {
```
Please make this a named function instead of a closure

pradipd replied:
Fixed in latest PR.

dnephin commented on integration/service/create_test.go:
```diff
@@ -0,0 +1,113 @@
+		return poll.Success()
+	})
+
+	err = client.NetworkRemove(context.Background(), overlayID)
```
Is this actually part of the test, or is it cleanup? If it's cleanup please move it out into a function with a descriptive name, and it should probably be done as a defer

pradipd replied:
Part of the test.

thaJeztah commented on daemon/cluster/executor/container/executor.go:
```diff
@@ -162,12 +173,17 @@ func (e *executor) Configure(ctx context.Context, node *api.Node) error {
+		return err
+	}
+
+	err = e.backend.AddLBAttachments(lbAttachments)
```
nit: this can be just 

```go
return e.backend.AddLBAttachments(lbAttachments)
```

pradipd replied:
Fixed in latest PR.

thaJeztah commented on daemon/network.go:
```diff
@@ -182,28 +182,7 @@ func (daemon *Daemon) setupIngress(create *clustertypes.NetworkCreateRequest, ip
-	if err := sb.EnableService(); err != nil {
-		logrus.Errorf("Failed enabling service for ingress sandbox")
-	}
+	daemon.createLoadBalancerSandbox("ingress", create.ID, ip, n, libnetwork.OptionIngress())
```
Why is the returned error not handled here?

pradipd replied:
I addressed this in the latest PR

thaJeztah commented on daemon/network.go:
```diff
@@ -235,6 +214,32 @@ func (daemon *Daemon) releaseIngress(id string) {
 
+//AddLBAttachments sets the mapping of LB IP address per network
+func (daemon *Daemon) AddLBAttachments(lbAttachments map[string]string) error {
+
```
Can you remove this empty line here (and other functions you added) to make the coding style match the existing code?

pradipd replied:
Done.

thaJeztah commented on daemon/network.go:
```diff
@@ -235,6 +214,32 @@ func (daemon *Daemon) releaseIngress(id string) {
+func (daemon *Daemon) AddLBAttachments(lbAttachments map[string]string) error {
+
+	for nid, nodeIP := range lbAttachments {
+
```
Same here, could you remove this empty line?

thaJeztah commented on daemon/cluster/executor/container/executor.go:
```diff
@@ -136,23 +136,34 @@ func (e *executor) Describe(ctx context.Context) (*api.NodeDescription, error) {
 }
 
 func (e *executor) Configure(ctx context.Context, node *api.Node) error {
-	na := node.Attachment
```
I'm not too familiar with this, but is `node.Attachment` no longer needed? (is this now included in `node.LbAttachments` ?)

pradipd replied:
Correct.  node.Attachment has been deprecated https://github.com/docker/swarmkit/blob/master/api/objects.proto#L64

thaJeztah reviewed (changes requested):
some nits and a question

fcrisciani commented on daemon/cluster/executor/container/executor.go:
```diff
@@ -136,23 +136,34 @@ func (e *executor) Describe(ctx context.Context) (*api.NodeDescription, error) {
+
+	if ingressNA == nil {
 		e.backend.ReleaseIngress()
+		e.backend.ClearLBAttachments()
```
is this function clearing the attachments of all the networks or only the Ingress?

pradipd replied:
It is clearing all the attachments for all networks --- which is wrong.
I fixed this in the latest PR.  
The lb attachments are always "reset".  I.e. they are cleared and then updated with the latest node.lbattachments passed in from the swarm manager.

fcrisciani commented on daemon/cluster/executor/container/executor.go:
```diff
@@ -162,12 +173,17 @@ func (e *executor) Configure(ctx context.Context, node *api.Node) error {
+		return err
+	}
+
+	err = e.backend.AddLBAttachments(lbAttachments)
```
The feeling here is that if there is not ingress network there is no lb configured, is it correct or there is a different code path followed?

pradipd replied:
If it is not ingress network, then we save the lb attachment (actually, only the per network IP address).
We only  use the lb IP address when we create a service.

fcrisciani commented on daemon/network.go:
```diff
@@ -496,6 +543,28 @@ func (daemon *Daemon) DeleteNetwork(networkID string) error {
 }
 
+func (daemon *Daemon) deleteLoadBalancerSandbox(n libnetwork.Network) {
+
```
I don't see a disableService here, wondering why there is an asymmetry between the EnableService in the Create and no need for the disableService here

pradipd replied:
Oops. I think I missed this in the last PR.  I'll update it.  sorry.

pradipd replied:
Address in the latest PR.

thaJeztah commented:
@pradipd vendor check is failing:

```
18:12:05 The result of vndr differs
18:12:05 
18:12:05  M vendor/github.com/docker/libnetwork/drivers/overlay/peerdb.go
18:12:05 
18:12:05 Please vendor your package with github.com/LK4D4/vndr.
```

pradipd commented:
Oops. I shouldn't have done s/sanbox/sandbox/ in the vendored packages.
will fix.

thaJeztah reviewed (approved):
thanks for addressing my nits; don't forget to squash commits to two commits; 1 for "vendoring" and 1 for the local changes. (We don't have to preserve the "fix vendor" and "address review comments" when merging)

pradipd committed:
vendoring libnetwork and swarmkit

Signed-off-by: Pradip Dhara <pradipd@microsoft.com>

dnephin commented on integration/service/create_test.go:
```diff
@@ -0,0 +1,129 @@
+	poll.WaitOn(t, serviceRunningTasksCount(client, serviceID, instances))
+
+	_, _, err = client.ServiceInspectWithRaw(context.Background(), serviceID, types.ServiceInspectOptions{})
+	require.NoError(t, err)
```
These 2 lines seem to be unnecessary. We already know the service can be inspected from the `poll.WaitOn()`. I think these 2 lines need to be removed.

pradipd replied:
Will remove in new PR.

dnephin commented on integration/service/create_test.go:
```diff
@@ -0,0 +1,129 @@
+	require.NoError(t, err)
+	assert.Contains(t, network.Containers, overlayName+"-sbox")
+
+	err = client.ServiceRemove(context.Background(), serviceID)
```
Everything from this line to the end of the test case seems to be test teardown, which should already be handled by `defer setupTest(t)()`. Why is all this necessary?

pradipd replied:
I want to validate that the network is removed cleanly.  
I see setupTest() is tearing everything down, but, it is forcefully shutting things down.  I want to make sure the typical user pattern works correctly.

dnephin commented on integration/service/create_test.go:
```diff
@@ -0,0 +1,129 @@
+	}
+}
+
+func serviceRunningTasksCount(client client.ServiceAPIClient, serviceID string, instances uint64) func(log poll.LogT) poll.Result {
```
This is a duplicate of `serviceContainerCount()` in inspect_test.go. I'm fine if you want to rename it, but I don't think we need both. We can just add the state check, which seems to be the only difference.

dnephin reviewed (commented):
Thank you for making those changes to the test case.

I think the test still needs a bit of cleanup. Would you mind opening a PR to address these points?

pradipd commented:
Will open up a new PR to address these.

pradipd commented:
https://github.com/moby/moby/pull/34898 is the PR.

stevvooe commented on daemon/cluster/executor/backend.go:
```diff
@@ -61,4 +62,5 @@ type Backend interface {
 	LookupImage(name string) (*types.ImageInspect, error)
 	PluginManager() *plugin.Manager
 	PluginGetter() *plugin.Store
+	GetLBAttachmentStore() *networkSettings.LBAttachmentStore
```
`GetAttachmentStore`

stevvooe commented on daemon/daemon.go:
```diff
@@ -121,6 +122,8 @@ type Daemon struct {
 	hosts            map[string]bool // hosts stores the addresses the daemon is listening on
 	startupDone      chan struct{}
+
+	lbAttachmentStore network.LBAttachmentStore
```
`attachmentStore`

stevvooe commented on daemon/network/settings.go:
```diff
@@ -31,3 +34,36 @@ type EndpointSettings struct {
 }
+
+// LBAttachmentStore stores the load balancer IP address for a network id.
+type LBAttachmentStore struct {
```
`AttachmentStore`.

stevvooe commented:
@pradipd I fixed the naming in swarmkit to remove LB. Could you please do so in moby, as well?

pradipd commented:
I have made the changes.  How do I sync with your changes in swarmkit?
Do I vendor in your branch/commit in vendor.conf?  
Or do I wait till you merge your changes into swarmkit, then vendor in the new master commit hash?

stevvooe commented:
@pradipd Wait till we merge the swarmkit changes and submit a PR with the updates.
